src/hotspot/share/code/codeBlob.cpp

```diff
@@ -63,42 +63,51 @@ unsigned int CodeBlob::align_code_offset(int offset) {
 
 // This must be consistent with the CodeBlob constructor's layout actions.
 unsigned int CodeBlob::allocation_size(CodeBuffer* cb, int header_size) {
-  unsigned int size = header_size;
-  size += align_up(cb->total_relocation_size(), oopSize);
   // align the size to CodeEntryAlignment
-  size = align_code_offset(size);
+  unsigned int size = align_code_offset(header_size);
   size += align_up(cb->total_content_size(), oopSize);
-  size += align_up(cb->total_oop_size(), oopSize);
-  size += align_up(cb->total_metadata_size(), oopSize);
   return size;
 }
 
 CodeBlob::CodeBlob(const char* name, CodeBlobKind kind, CodeBuffer* cb, int size, uint16_t header_size,
-                   int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments) :
+                   int16_t frame_complete_offset, int frame_size, OopMapSet* oop_maps, bool caller_must_gc_arguments,
+                   int mutable_data_size) :
   _oop_maps(nullptr), // will be set by set_oop_maps() call
   _name(name),
   _size(size),
   _relocation_size(align_up(cb->total_relocation_size(), oopSize)),
-  _content_offset(CodeBlob::align_code_offset(header_size + _relocation_size)),
+  _content_offset(CodeBlob::align_code_offset(header_size)),
   _code_offset(_content_offset + cb->total_offset_of(cb->insts())),
-  _data_offset(_content_offset + align_up(cb->total_content_size(), oopSize)),
   _frame_size(frame_size),
   S390_ONLY(_ctable_offset(0) COMMA)
   _header_size(header_size),
   _frame_complete_offset(frame_complete_offset),
   _kind(kind),
-  _caller_must_gc_arguments(caller_must_gc_arguments)
+  _caller_must_gc_arguments(caller_must_gc_arguments),
+  _mutable_data(nullptr),
+  _mutable_data_size(0)
 {
   assert(is_aligned(_size,            oopSize), "unaligned size");
   assert(is_aligned(header_size,      oopSize), "unaligned size");
   assert(is_aligned(_relocation_size, oopSize), "unaligned size");
-  assert(_data_offset <= _size, "codeBlob is too small: %d > %d", _data_offset, _size);
+  int code_end_offset = _content_offset + align_up(cb->total_content_size(), oopSize);
+  assert(code_end_offset == _size, "wrong codeBlob size: %d != %d", _size, code_end_offset);
   assert(code_end() == content_end(), "must be the same - see code_end()");
 #ifdef COMPILER1
   // probably wrong for tiered
   assert(_frame_size >= -1, "must use frame size or -1 for runtime stubs");
 #endif // COMPILER1
 
+  // The mutable_data_size is either calculated by the nmethod constructor to account
+  // for reloc_info and additional data, or it is set here to accommodate only the relocation data.
+  _mutable_data_size = (mutable_data_size == 0) ? cb->total_relocation_size() : mutable_data_size;
```

Comment:
This seems strange to treat relocations as special.  Wouldn't it be better to have the caller always pass in the correct value?

Reply:
Or compute using something like required_mutable_data_space()?

Reply:
Alright. Thank you. I moved mutable_data_size calculation out of CodeBlob.